### PR TITLE
refactor(io_expander): support i2c ng

### PR DESCRIPTION
**Pull request by Lzw655**
# ESP-BSP Pull Request checklist

> Note: For new BSPs create a PR with this [link](https://github.com/espressif/esp-bsp/compare/main...my-branch?quick_pull=1&template=pr_template_bsp.md).
- [x] Version of modified component bumped
- [x] CI passing

# Change description
_Please describe your change here_

### CODE REVIEW

**Comment by Lzw655** [comment]
Hi @tore-espressif @espzav, PTAL, thank you!

**Comment by tore-espressif** [comment]
@Lzw655 I rebased your branch so it gets some fixes from master. I'll have a look ASAP, thank you!

File: .build-test-rules.yml [hunk 1]
```diff
@@ -84,3 +84,24 @@ components/icm42670:
 components/qma6100p:
   depends_filepatterns:
     - "components/qma6100p/**"
+
+components/io_expander/esp_io_expander_ht8574:
+  depends_filepatterns:
+    - "components/io_expander/esp_io_expander_ht8574/**"
```
Comment (tore-espressif) [inline .build-test-rules.yml:90]: The component also depends on `components/io_expander/esp_io_expander`. Could you please add it to all 3 test_apps?

File: components/io_expander/esp_io_expander_tca95xx_16bit/test_apps/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# The following lines of boilerplate have to be in your project's CMakeLists
+# in this exact order for cmake to work correctly
+cmake_minimum_required(VERSION 3.5)
+set(EXTRA_COMPONENT_DIRS "$ENV{IDF_PATH}/tools/unit-test-app/components")
```
Comment (tore-espressif) [inline components/io_expander/esp_io_expander_tca95xx_16bit/test_apps/CMakeLists.txt:4]: I don't see any function from `$ENV{IDF_PATH}/tools/unit-test-app/components/test_utils` in this test. Can it be removed (in all 3 test_apps)?

File: components/io_expander/esp_io_expander_ht8574/esp_io_expander_ht8574.c [hunk 3]
```diff
@@ -83,13 +92,9 @@ esp_err_t esp_io_expander_new_i2c_ht8574(i2c_port_t i2c_num, uint32_t i2c_addres
...
     esp_io_expander_ht8574_t *ht8574 = (esp_io_expander_ht8574_t *)__containerof(handle, esp_io_expander_ht8574_t, base);
-
     uint8_t temp = 0;
-    // *INDENT-OFF*
-    ESP_RETURN_ON_ERROR(
-        i2c_master_read_from_device(ht8574->i2c_num, ht8574->i2c_address, &temp, 1, pdMS_TO_TICKS(I2C_TIMEOUT_MS)),
-        TAG, "Read input reg failed");
-    // *INDENT-ON*
+
+    ESP_RETURN_ON_ERROR(i2c_master_receive(ht8574->i2c_handle, &temp, sizeof(temp), -1), TAG, "Read input reg failed");
```
Comment (tore-espressif) [inline components/io_expander/esp_io_expander_ht8574/esp_io_expander_ht8574.c:97]: Could we leave the timeout here? Also in other places and in all 3 drivers
```suggestion
    ESP_RETURN_ON_ERROR(i2c_master_receive(ht8574->i2c_handle, &temp, sizeof(temp), I2C_TIMEOUT_MS), TAG, "Read input reg failed");
```

**Review by tore-espressif — APPROVED** [review]
LGTM @Lzw655 ! I left few nitpick comments and 1 important about I2C commands timeout

**Comment by Lzw655** [comment]
@tore-espressif Thank you for reviewing! All fixed.

File: components/io_expander/esp_io_expander_ht8574/esp_io_expander_ht8574.c [hunk 4]
```diff
@@ -1,23 +1,22 @@
...
+ * SPDX-FileCopyrightText: 2023-2024 Espressif Systems (Shanghai) CO LTD
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #include <inttypes.h>
 #include <string.h>
 #include <stdlib.h>
-
 #include "driver/i2c.h"
```
Comment (tore-espressif) [inline components/io_expander/esp_io_expander_ht8574/esp_io_expander_ht8574.c:10]: I'm sorry, one last thing. Please remove this deprecated header

**Comment by tore-espressif** [comment]
> @tore-espressif Thank you for reviewing! All fixed.

@Lzw655 Thank you for the quick update, I have one last comment and we can merge

File: components/io_expander/esp_io_expander_ht8574/esp_io_expander_ht8574.c (hunk 4 above)
Comment (Lzw655) [reply components/io_expander/esp_io_expander_ht8574/esp_io_expander_ht8574.c]: Sorry for overlooking this, it’s now fixed.

**Pull request merged**